Add support for Under Filesystem eventual consistency issue

LuQQiu opened the pull request:
#8426

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2545/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2551/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2552/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2553/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2554/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2572/
Test PASSed.

calvinjia commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -231,4 +318,59 @@ public void stopSync(AlluxioURI uri) { }
+      try {
+        T result  = op.apply();
+        if (result instanceof Boolean) {
+          if ((Boolean) result) {
```
This is kind of questionable, does `false` always imply a failure?

LuQQiu replied:
Changed to use `RetryOnFailure` and `RetryOnFalse`, `false` in our `BaseFileSystem` means failures like `unable to rename File/Directory`, `Fail to delete directory`.

calvinjia commented on core/common/src/main/java/alluxio/underfs/ConsistentUnderFileSystem.java:
```diff
@@ -0,0 +1,112 @@
+
+/**
+ * Representing a consistent under filesystem
+ * which does not have eventual consistency issue.
```
Formatting

LuQQiu replied:
Changed to 	 
```
* Represents a consistent under filesystem
 * which does not have eventual consistency issues.
```

LuQQiu replied:
Changed, thanks

aaudiber commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -2772,7 +2772,7 @@ private void mountInternal(Supplier<JournalContext> journalContext, LockedInodeP
                 ServerConfiguration.global()));
         // Check that the ufsPath exists and is a directory
-        if (!ufs.isDirectory(ufsPath.toString())) {
+        if (!ufs.isExistingDirectory(ufsPath.toString())) {
```
we don't know that it's existing at this point, so we should use regular `isDirectory`

LuQQiu replied:
Changed, thanks

aaudiber replied:
agree, I think we can remove `isExistingDirectory`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2643/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2667/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2668/
Test FAILed.

LuQQiu commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2676/
Test PASSed.

LuQQiu commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -82,6 +85,35 @@ public boolean deleteDirectory(String path) throws IOException {
+
+  @Override
+  public OutputStream createNonexistingFile(String path, CreateOptions options) throws IOException {
+    return retry(() -> createNonexistingFile(path, options),
```
Changed, thanks

LuQQiu commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -82,6 +85,35 @@ public boolean deleteDirectory(String path) throws IOException {
+  @Override
+  public OutputStream createNonexistingFile(String path) throws IOException {
+    return retry(() -> createNonexistingFile(path, CreateOptions.defaults(mAlluxioConf)),
+        "create file " + path);
```
No sure if it is the right way:
I changed to retry functions to take `Supplier<String> description` as input
```
retryOnFalse(UfsBooleanOperation<Boolean> op, Supplier<String> description)
```
When catch exception, I called `description.get()`.

By this way, I can pass in the closure format
```
  @Override
  public OutputStream createNonexistingFile(String path) throws IOException {
    return retryOnException(() -> create(path, CreateOptions.defaults(mAlluxioConf)),
        () -> "create file " + path);
  }
```

LuQQiu commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -231,4 +318,59 @@ public void stopSync(AlluxioURI uri) { }
+      try {
+        T result  = op.apply();
+        if (result instanceof Boolean) {
+          if ((Boolean) result) {
```
Changed to `retryOnException` and `retryOnFalse`, Much clear now, thanks

LuQQiu commented on core/server/common/src/main/java/alluxio/cli/validation/UfsDirectoryValidationTask.java:
```diff
@@ -38,7 +38,7 @@ public UfsDirectoryValidationTask() {
   public TaskResult validate(Map<String, String> optionsMap) {
     try {
-      UfsStatus[] listStatus = mUfs.listStatus(mPath);
+      UfsStatus[] listStatus = mUfs.listExistingStatus(mPath);
```
Changed.
I did not change the underfilesystem calls in the journal codes because of the same reason.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -175,7 +175,7 @@ private void initFromBackup(AlluxioURI backup) throws IOException {
     }
     try (UnderFileSystem closeUfs = ufs;
-        InputStream ufsIn = ufs.open(backup.getPath())) {
+        InputStream ufsIn = ufs.openExistingFile(backup.getPath())) {
```
Changed, thanks

LuQQiu commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -1197,7 +1197,7 @@ private boolean checkConsistencyInternal(LockedInodePath inodePath) throws Inval
       } else {
         return ufs.isFile(ufsPath)
-            && ufs.getFileStatus(ufsPath).getContentLength() == inode.asFile().getLength();
+            && ufs.getExistingFileStatus(ufsPath).getContentLength() == inode.asFile().getLength();
```
Changed to `getFileStatus` because if the `ufs.isFile` (a list operation) can find the file, we should be able to get file status (another list operation).

LuQQiu commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -2214,7 +2214,7 @@ private void renameInternal(RpcContext rpcContext, LockedInodePath srcInodePath,
         mUfsAbsentPathCache.processExisting(dstPath);
       }
-    } catch (Exception e) {
+    } catch (IOException e) {
```
Changed to `Throwable`. Thanks

LuQQiu commented:
@aaudiber 
Main Changes:
1. Added the integration tests in the `AbstractUnderFileSystemContractTest` and tested against S3.
2. Remove the `listExistingStatus`.
After we create directory, we use `listExistingStatus` will not help us to gain the latest written files in the directory. `listExistingStatus` give us the exact same results as `listStatus`.
After we create a file, `listStatus` on this file will return null in the `ObjectUnderFileSystem`.
Retry on null is strange.

Addressed the other comments, PTAL, thanks

aaudiber commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -231,4 +310,91 @@ public void stopSync(AlluxioURI uri) { }
+  /**
+   * Represents a under filesystem operation with a boolean return type.
+   */
+  private interface UfsBooleanOperation<Boolean> {
```
no need for the `Boolean` generic argument - it doesn't get used anywhere

LuQQiu replied:
Removed the boolean, `retryOnException` and `retryOnFalse` both use the `ObjectStoreOperation<T>`.

aaudiber commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -231,4 +310,91 @@ public void stopSync(AlluxioURI uri) { }
+   * @param description the description regarding the operation
+   * @return the operation result if operation succeed or returned true
+   */
+  private boolean retryOnFalse(UfsBooleanOperation<Boolean> op,
```
can we write a helper method to combine the common logic between `retryOnException` and `retryOnFalse`? The only difference is what conditions they retry under.

I think the signature could be something like

```java
private <T> retry(Callable<T> op, BiFunction<Throwable, T, Boolean> shouldRetry)
```

LuQQiu replied:
The `retryOnFalse` I have changed to only retry on false. Now the two functions have more different places, would you think it is still helpful to combine the logics?
```
  private boolean retryOnFalse(ObjectStoreOperation<Boolean> op,
      Supplier<String> description) throws IOException {
    RetryPolicy retryPolicy = getRetryPolicy();
    while (retryPolicy.attempt()) {
      if (op.apply()) {
        return true;
      } else {
        LOG.debug("Failed in attempt {} to {} ", retryPolicy.getAttemptCount(),
            description.get());
      }
    }
    return false;
  }

  private <T> T retryOnException(ObjectStoreOperation<T> op,
      Supplier<String> description) throws IOException {
    RetryPolicy retryPolicy = getRetryPolicy();
    IOException thrownException = null;
    while (retryPolicy.attempt()) {
      try {
        return op.apply();
      } catch (IOException e) {
        LOG.debug("Attempt {} to {} failed with exception : {}", retryPolicy.getAttemptCount(),
            description.get(), e.toString());
        thrownException = e;
      }
    }
    throw thrownException;
  }
```

aaudiber replied:
I think it would be a small improvement, but if you prefer two methods I'm fine keeping it how it is now.

FYI you can get java syntax highlighting in code snippets by starting the block with "```java"

aaudiber commented on core/common/src/main/java/alluxio/conf/PropertyKey.java:
```diff
@@ -611,30 +611,30 @@ public String toString() {
-          .setDescription("Block reads from an object store automatically retry for transient "
-              + "errors with an exponential backoff. This property determines the base time in the "
-              + "exponential backoff. Only applicable for S3A.")
+          .setDescription("Block eventual-consistency-related operations "
```
"To handle eventually consistent storage semantics for certain under storages, Alluxio will perform retries when under storage metadata doesn't match Alluxio's expectations. These retries use exponential backoff. This property determines the base time for the exponential backoff."

LuQQiu replied:
Changed, thanks

aaudiber commented on core/common/src/test/java/alluxio/underfs/AbstractUnderFileSystemContractTest.java:
```diff
@@ -626,6 +682,16 @@ public void renameFile() throws IOException {
   }
 
+  @Test
+  public void RenameRenamableFile() throws IOException {
```
Rename -> rename

LuQQiu replied:
Changed, thanks

aaudiber commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -82,6 +86,35 @@ public boolean deleteDirectory(String path) throws IOException {
 
+  @Override
+  public OutputStream createNonexistingFile(String path) throws IOException {
+    return retryOnException(() -> create(path, CreateOptions.defaults(mAlluxioConf)),
```
what do you think of moving the default retrying implementations to `ObjectUnderFileSystem`? Then we don't need to override existing implementations in `ConsistentUnderFileSystem`. If we add a new retry-style method to `UnderFileSystem`, it would be nice if the compiler forced us to add a consistent implementation to `ConsistentUnderFileSystem`. Right now, we could easily forget to override a retry-style method in `ConsistentUnderFileSystem`, and end up doing retries there.

LuQQiu replied:
Agree! by now the eventual consistency issue only occur in the Object storages. if in the future, other kind of storages also have this issue, we can revert to implement retry in `BaseUnderFileSystem` and overwrite in `ConsistentUnderFileSystem`.

aaudiber replied:
My general point is that I don't like having default implementations in `BaseFileSystem`. That way, if we add a new method in `BaseFileSystem`, we are forced by the compiler to provide appropriate implementations in both `ConsistentUnderFileSystem` and `ObjectUnderFileSystem`. Right now, if we add a new method to `BaseUnderFileSystem`, we also need to remember to override it in `ObjectUnderFileSystem`. If we forget, `ObjectUnderFileSystem` will use the default behavior, and we could have consistency issues again.

LuQQiu replied:
Oh, I see. Do not give default retry in `BaseFileSystem` to force us implement `consistent ways` in `ConsistentUnderFileSystem` and deal with eventual consistency in `objectUnderFileSystem`. Thanks!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2767/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2774/
Test PASSed.

LuQQiu commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -231,4 +310,91 @@ public void stopSync(AlluxioURI uri) { }
+        return op.apply();
+      } catch (IOException e) {
+        LOG.debug("{} attempt to {} failed with exception : {}", retryPolicy.getAttemptCount(),
+            description, e.getMessage());
```
Changed, thanks

LuQQiu commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -231,4 +310,91 @@ public void stopSync(AlluxioURI uri) { }
+        return op.apply();
+      } catch (IOException e) {
+        LOG.debug("{} attempt to {} failed with exception : {}", retryPolicy.getAttemptCount(),
+            description, e.getMessage());
```
Changed to the `e.toString()`, the exception class name is also helpful. thanks

LuQQiu commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -231,4 +310,91 @@ public void stopSync(AlluxioURI uri) { }
+      try {
+        return op.apply();
+      } catch (IOException e) {
+        LOG.debug("{} attempt to {} failed with exception : {}", retryPolicy.getAttemptCount(),
```
Changed to `Attempt {}`, sounds much better, thanks

LuQQiu commented on core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:
```diff
@@ -231,4 +310,91 @@ public void stopSync(AlluxioURI uri) { }
+      } catch (IOException e) {
+        LOG.debug("{} attempt to {} failed with exception : {}", retryPolicy.getAttemptCount(),
+            description.get(), e.getMessage());
+        thrownException = e;
```
Changed to only retry on false. I saw those under filesystem operations that returns boolean do not throw exceptions. When they have exceptions, they catch them and return false.

LuQQiu commented:
@aaudiber Addressed the comments, please take another round of review, thanks

aaudiber commented:
@LuQQiu I left one last comment, otherwise looks good to me

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/2777/
Test PASSed.

LuQQiu commented:
@aaudiber Removed the default implementations from `BaseUnderFileSystem` to force overwrite in `ConsistentUnderFileSystem` and `ObjectUnderFileSystem`. PTAL, thanks

LuQQiu commented:
Thanks @aaudiber 
@calvinjia PTAL, thanks

calvinjia commented:
alluxio-bot, merge this please.
